tools/syz-declextract/syz-declextract.cpp

```diff
@@ -162,41 +171,163 @@ class Matcher : public MatchFinder::MatchCallback {
+      // TODO: Extract extern variables
+      return {};
+    }
+    const auto *opsInit = llvm::dyn_cast<InitListExpr>(opsDecl->getInit());
+    for (const auto &field : opsInit->getInit(0)->getType()->getAsRecordDecl()->fields()) {
+      opsMember[opsName][field->getNameAsString()] = field->getFieldIndex();
+    }
+    for (int i = 0; i < n_ops; ++i) {
+      const auto &init = llvm::dyn_cast<InitListExpr>(opsInit->getInit(i));
+      const auto &cmdInit = init->getInit(opsMember[opsName]["cmd"])->getEnumConstantDecl();
+      if (!cmdInit) {
+        continue;
+      }
+      const auto &cmd = cmdInit->getNameAsString();
+      const ValueDecl *policyDecl = nullptr;
+      if (opsName != "small_ops") {
+        policyDecl = init->getInit(opsMember[opsName]["policy"])->getAsBuiltinConstantDeclRef(*context);
+      }
+      cmds.push_back({std::move(cmd), getPolicyName(Result, policyDecl)});
+    }
+    return cmds;
+  }
+
+  void genlFamily(const MatchFinder::MatchResult &Result) {
+    ASTContext *context = Result.Context;
+    const auto *genlFamily = Result.Nodes.getNodeAs<RecordDecl>("genl_family");
+    if (!genlFamily) {
+      return;
+    }
+    for (const auto &field : genlFamily->fields()) {
+      genlFamilyMember[field->getNameAsString()] = field->getFieldIndex();
+    }
+    const auto *genlFamilyInit = Result.Nodes.getNodeAs<InitListExpr>("genl_family_init");
+    if (!genlFamilyInit) {
+      return;
+    }
+
+    auto name = llvm::dyn_cast<StringLiteral>(genlFamilyInit->getInit(genlFamilyMember["name"]))->getString().str();
+    const auto &globalPolicyName =
+        genlFamilyInit->getInit(genlFamilyMember["policy"])->getAsBuiltinConstantDeclRef(*context);
+
+    std::string familyPolicyName;
+    if (globalPolicyName) {
+      familyPolicyName = getPolicyName(Result, globalPolicyName).value();
+    }
+
+    std::replace(name.begin(), name.end(), '.', '_'); // Illegal character.
+    std::replace(name.begin(), name.end(), ' ', '_'); // Don't leave space in name.
+    char msghdr[100];
+    sprintf(msghdr, "msghdr_%s_auto", name.c_str());
+    bool printedCmds = false;
+    for (const auto &opsType : {"ops", "small_ops", "split_ops"}) {
+      for (auto &ops : getOps(Result, opsType, genlFamilyInit)) {
+        const char *policyName;
+        if (ops.policy.has_value()) {
+          policyName = ops.policy.value().c_str();
+        } else if (globalPolicyName) {
+          policyName = familyPolicyName.c_str();
+        } else {
+          continue;
+        }
+        printf("sendmsg$auto_%s(fd sock_nl_generic, msg ptr[in, %s[%s, %s]], f flags[send_flags]) (automatic)\n",
+               ops.cmd.c_str(), msghdr, ops.cmd.c_str(), policyName);
+        printedCmds = true;
+      }
+    }
+    if (!printedCmds) { // Do not print resources and types if they're not used in any cmds
+      return;
+    }
+    char resourceName[100];
+    sprintf(resourceName, "genl_%s_family_id_auto", name.c_str());
```

Comment:
This can silently corrupt memory. At least use snprintf, or something else.

Reply:
I used `std::string` instead!